src/hotspot/share/logging/logAsyncFlusher.hpp

```diff
@@ -0,0 +1,169 @@
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+#ifndef SHARE_LOGGING_ASYNC_FLUSHER_HPP
+#define SHARE_LOGGING_ASYNC_FLUSHER_HPP
+#include "logging/log.hpp"
+#include "logging/logDecorations.hpp"
+#include "logging/logFileOutput.hpp"
+#include "logging/logMessageBuffer.hpp"
+#include "memory/resourceArea.hpp"
+#include "runtime/mutexLocker.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "utilities/hashtable.hpp"
+#include "utilities/linkedlist.hpp"
+
+template <typename E, MEMFLAGS F>
+class LinkedListDeque : private LinkedListImpl<E, ResourceObj::C_HEAP, F> {
+ private:
+  LinkedListNode<E>* _tail;
+  size_t _size;
+
+ public:
+  LinkedListDeque() : _tail(NULL), _size(0) {}
+  void push_back(const E& e) {
+    if (!_tail) {
+      _tail = this->add(e);
+    } else {
+      _tail = this->insert_after(e, _tail);
+    }
+
+    ++_size;
+  }
+
+  void pop_all(LinkedList<E>* logs) {
+    logs->move(static_cast<LinkedList<E>* >(this));
+    _tail = NULL;
+    _size = 0;
+  }
+
+  void pop_front() {
+    LinkedListNode<E>* h = this->unlink_head();
+    if (h == _tail) {
+      _tail = NULL;
+    }
+
+    if (h != NULL) {
+      --_size;
+      this->delete_node(h);
+    }
+  }
+
+  size_t size() const { return _size; }
+
+  const E* front() const {
+    return this->_head == NULL ? NULL : this->_head->peek();
+  }
+
+  const E* back() const {
+    return _tail == NULL ? NULL : _tail->peek();
+  }
+};
+
+class AsyncLogMessage {
+  LogFileOutput&   _output;
+  LogDecorations*  _decorations;
+  char*            _message;
+
+public:
+  AsyncLogMessage(LogFileOutput& output, const LogDecorations& decorations, const char* msg)
+    : _output(output) {
+    _decorations = new LogDecorations(decorations);
+      // allow to fail here, then _message is NULL
+    _message = os::strdup(msg, mtLogging);
+  }
+
+  void destroy() {
+    delete _decorations;
+    _decorations = NULL;
+
+    if (_message != NULL) {
+      os::free(_message);
+    }
+  }
+
+  void writeback();
+
+  // two AsyncLogMessage are equal if both _output and _message are same.
+  bool equals(const AsyncLogMessage& o) const {
+    if (_message == o._message) {
+      return &_output == &o._output;
+    } else if (_message == NULL || o._message == NULL) {
+      return false;
+    } else {
+      return &_output == &o._output && !strcmp(_message, o._message);
+    }
+  }
+
+  const char* message() const { return _message; }
+  LogFileOutput* output() const { return &_output; }
+};
+
+typedef LinkedListDeque<AsyncLogMessage, mtLogging> AsyncLogBuffer;
+typedef KVHashtable<LogFileOutput*, uintx, mtLogging> AsyncLogMap;
```

Comment:
This is just to keep statistics of dropped messages per output, right? 

I think it would be a lot simpler just to add a reset-able counter to LogFileOutput directly. I don't even think it has to be uintx. 32bit should be sufficient (also, uintx makes no sense - do we somehow expect less message drops on 32bit platforms?)

Reply:
Yes, it is for the statistics of dropped messages.

AsyncLogMap looks a little cumbersome indeed, but the good side is that I isolate async logging things in logAsyncFlusher.hpp/cpp.  I manage not to pollute shared unified logging code. 

Another reason is that I can deliver the zero-cost promise when async logging is off.  If I move the counter to LogFileOutput, it will increase those object size a little bit.  In current implementation,  the singleton object LogAsyncFlusher is not created if -Xlog:async is absent.

Size wise, I think it's okay to to have 8 byte.  if a network-based device disappeared, who know how long asynclog thread  would block.  bigger counter is better.  Cost is almost same because the number of logFileOutput objects is usually less than 4.

Reply:
I respect your "no costs promise" concern, but really this is no problem. We create exactly one LogFileOutput object per file sink; so usually we have one, maybe two of those if async=true. I think the costs of those counters would be bearable :)

I understand the concern of not polluting the UL code base. But here I think it does not apply, since you bring your patch upstream anyway, so async logging is now part of the UL code base.

wrt to counter size, okay lets use 64bit. But lets really use 64bit, also in 32bit VMs (so, use uint64_t instead of uintx).

Reply:
Hi, @tstuefe , 
I agree with you that uintx is not a good type. uint32_t is big enough. 

I found it's  very tricky to move those counters to LogFileOutput objects. after than, I still need HashTable as a set to traverse all LogFileObject objects. 

Alternatively, there's a member variable `_outputs` in LogConfiguration, which refers to all LogOutput objects. I have to obtain  `ConfigurationLock` first before reading it because jcmd could modify outputs at any moment.  Grabbing `ConfigurationLock`  in `LogAsyncFlusher::flush` may lead to a dead lock with `_lock`.  This part of code is really subtle. I tried different approaches. no one is  more straight-forward than a hashmap.